### PR TITLE
Added header fields validation against not allowed characters

### PR DESCRIPTION
**Pull request by adrianosferreira**
Reported on https://github.com/guzzle/guzzle/issues/1870

Whitelisted / blacklisted a bunch of characters based on https://tools.ietf.org/html/rfc7230#section-3.2

### CODE REVIEW

**Comment by Tobion** [comment]
Please use 

```
if (! preg_match('/^[a-zA-Z0-9\'`#$%&*+.^_|~!-]+$/', $name)) {
            throw new Exception\InvalidArgumentException(sprintf(
                '"%s" is not valid header name',
                $name
            ));
        }
```

referring to https://tools.ietf.org/html/rfc7230#section-3.2

**Comment by adrianosferreira** [comment]
@Tobion done, please double-check.

File: src/MessageTrait.php [hunk 1]
```diff
@@ -212,5 +212,14 @@ private function assertHeader($header)
         if ($header === '') {
             throw new \InvalidArgumentException('Header name can not be empty.');
```
Comment (Tobion) [inline src/MessageTrait.php]: This is not needed anymore given the regex. Please add empty string to the tests cases.

File: tests/RequestTest.php [hunk 2]
```diff
@@ -185,6 +185,68 @@ public function testCanGetHeaderAsCsv()
         self::assertEquals('', $r->getHeaderLine('Bar'));
     }
 
+    /**
+     * @dataProvider dpHeadersContainingNotAllowedChars
```
Comment (Tobion) [inline tests/RequestTest.php]: please replace "dp" with "provide" for better readability

File: tests/RequestTest.php [hunk 3]
```diff
@@ -185,6 +185,68 @@ public function testCanGetHeaderAsCsv()
...
+        );
+    }
+
+    public function dpHeadersContainingNotAllowedChars()
+    {
+        return [[' key '], ['key '], [' key'], ['key/'], ['key('], ['key\\']];
+    }
+
+    /**
+     * @dataProvider dpHeadersContainsAllowedChar
```
Comment (Tobion) [inline tests/RequestTest.php]: same "provide"

File: src/MessageTrait.php [hunk 4]
```diff
@@ -212,5 +212,14 @@ private function assertHeader($header)
...
         }
+
+        if (! preg_match('/^[a-zA-Z0-9\'`#$%&*+.^_|~!-]+$/', $header)) {
```
Comment (Tobion) [inline src/MessageTrait.php]: please link to https://tools.ietf.org/html/rfc7230#section-3.2 in the `assertHeader` method phpdoc

**Comment by adrianosferreira** [comment]
@Tobion changes done, please double-check once more.

**Comment by Tobion** [comment]
May I ask you to rebase again on the latest master as the phpstan error has been fixed there.

**Comment by adrianosferreira** [comment]
@Tobion done, indeed, it fixed the issue.

**Comment by Tobion** [comment]
Thank you @adrianosferreira

**Pull request merged**